src/Kestrel.Core/Internal/Http/HttpProtocol.cs

```diff
@@ -615,6 +483,156 @@ public void OnHeader(Span<byte> name, Span<byte> value)
+                try
+                {
+                    KestrelEventSource.Log.RequestStart(this);
+
+                    // Run the application code for this request
+                    await application.ProcessRequestAsync(httpContext);
+
+                    if (_requestAborted == 0)
+                    {
+                        VerifyResponseContentLength();
+                    }
+                }
+                catch (Exception ex)
+                {
+                    ReportApplicationError(ex);
+                    // Capture BadHttpRequestException for further processing
+                    badRequestException = ex as BadHttpRequestException;
+                }
+
+                KestrelEventSource.Log.RequestStop(this);
+
+                // Trigger OnStarting if it hasn't been called yet and the app hasn't
+                // already failed. If an OnStarting callback throws we can go through
+                // our normal error handling in ProduceEnd.
+                // https://github.com/aspnet/KestrelHttpServer/issues/43
+                if (!HasResponseStarted && _applicationException == null && _onStarting != null)
+                {
+                    await FireOnStarting();
+                }
+
+                PauseStreams();
+
+                if (_onCompleted != null)
+                {
+                    await FireOnCompleted();
+                }
+
+                if (badRequestException == null)
+                {
+                    // If _requestAbort is set, the connection has already been closed.
+                    if (_requestAborted == 0)
+                    {
+                        // Call ProduceEnd() before consuming the rest of the request body to prevent
+                        // delaying clients waiting for the chunk terminator:
+                        //
+                        // https://github.com/dotnet/corefx/issues/17330#issuecomment-288248663
+                        //
+                        // This also prevents the 100 Continue response from being sent if the app
+                        // never tried to read the body.
+                        // https://github.com/aspnet/KestrelHttpServer/issues/2102
+                        //
+                        // ProduceEnd() must be called before _application.DisposeContext(), to ensure
+                        // HttpContext.Response.StatusCode is correctly set when
+                        // IHttpContextFactory.Dispose(HttpContext) is called.
+                        await ProduceEnd();
+
+                        // ForZeroContentLength does not complete the reader nor the writer
+                        if (!messageBody.IsEmpty)
+                        {
+                            try
+                            {
+                                // Finish reading the request body in case the app did not.
+                                await messageBody.ConsumeAsync();
```

Comment:
How confident are we that this is the only thing that can throw after invoking the application and verifying the response body length?

Personally, I'm fairly confident this is the case, but I want to make extra sure that we dispose the HttpContext in all the common scenarios.

Reply:
Was thinking `FireOnStarting` and `FireOnCompleted` if the user code the executed threw; but they don't let exceptions escape.

Reply:
Yep. I think this change is safe. I just wanted an extra pair of eyes.